File: Sean的leetcode/Leetcode/Linked List Cycle II.cpp
//refer to the solution of : http://www.cnblogs.com/hiddenfox/p/3408931.html
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode *slow = head, *fast = head;
        ListNode *curr;

        if (!head) return 0;

        while (slow && fast)
        {
            slow = slow->next;
            if (fast->next)
                fast = fast->next->next;
            else
                return 0;

            if (slow == fast)
                break;
        }

        if (!slow || !fast)
            return 0;

        //the distance between the head and cycle beginning is the same as
        //the distance between the meeting point and the cycle beginning :  a: head to loop start, b: loop start to meeting point, x: meeting point to loop start,slow: a+b, fast: a+b+x+b = 2(a+b) => x = a
        curr = head;
        while (curr != slow)
        {
            curr = curr->next;
            slow = slow->next;
        }
        return curr;
    }
};

File: MyLeetCode/python/Linked List Cycle II.py
"""
Linked list - slow & fast pointers
Time Complexity: O(n)
Space Complexity: O(1)
Break the cycle to two part: a - both slow and fast passed by
                             b - only fast passed by
And the length outside the cycle is L
steps(slow) = L + a
steps(fast) = L + a + b + a = 2 (L + a) => b = L
So we will
(1) find the intersect
(2) use two pointers to start from head and intersect until they meet
"""
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:      
        intersect = self.getIntersect(head)
        if not intersect:
            return None
        ptr1 = head
        ptr2 = intersect
        while ptr1 != ptr2:
            ptr1 = ptr1.next
            ptr2 = ptr2.next
        return ptr1
        
    def getIntersect(self, head):   
        f = s = head
        while f and f.next:
            f = f.next.next
            s = s.next
            if s == f:
                return s.next
        return None
        